File: packages/openland-graphql/spacex/store/selectorKey.ts
import { InputValue } from './../types';

function selectorValueKey(src: any): string {
    if (src === null) {
        return 'null';
    } else if (typeof src === 'string') {
        return '"' + src + '"';
    } else if (Array.isArray(src)) {
        return '[' + src.map((v) => selectorValueKey(v)).join(',') + ']';
    } else if (typeof src === 'object') {
        let keys = Object.keys(src);
        keys.sort();
        let vals: string[] = [];
        for (let k of keys) {
            vals.push(k + ':' + selectorValueKey(src[k]));
        }
        return '{' + vals.join(',') + '}';
    } else {
        return '' + src;
    }
}

function selectorInputValueKey(value: InputValue, queryArguments: any): string | null {
    if (value.type === 'int') {
        return value.value + ''; // Check type?
    } else if (value.type === 'float') {
        return value.value + ''; // Check type?        
    } else if (value.type === 'boolean') {
        return value.value + ''; // Check type?
    } else if (value.type === 'string') {
        return `"${value.value}"`;
    } else if (value.type === 'null') {
        return 'null';
    } else if (value.type === 'list') {
        return '[' + value.items.map((v) => selectorInputValueKey(v, queryArguments)).join(',') + ']';
    } else if (value.type === 'object') {

        // Read and sort all keys
        let keys = Object.keys(value.fields);
        keys.sort();

        // Calculate selector keys
        let fields: string[] = [];
        for (let k of keys) {
            let sk = selectorInputValueKey(value.fields[k] as InputValue, queryArguments);
            if (sk !== null /* Do not reduce condition */) {
                fields.push(k + ':' + sk);
            }
        }

        // Compute result
        return '{' + fields.join(',') + '}';
    } else if (value.type === 'reference') {
        let name = value.name;
        if (queryArguments[name] !== undefined) {
            return selectorValueKey(queryArguments[name]);
        } else {
            return null;
        }
    } else {
        throw Error('Unknown InputValue');
    }
}

export function selectorKey(name: string, fieldArguments: { [key: string]: InputValue }, queryArguments: any) {
    if (Object.keys(fieldArguments).length === 0) {
        return name;
    }
    let keys = Object.keys(fieldArguments);
    keys.sort();
    let values: string[] = [];
    for (let k of keys) {
        let sk = selectorInputValueKey(fieldArguments[k] as InputValue, queryArguments);
        if (sk !== null) {
            values.push(k + ':' + sk);
        }
    }
    if (values.length === 0) {
        return name;
    }
    return name + '(' + values.join(',') + ')';
}

File: android/app/src/main/java/com/openland/spacex/store/selectorKey.kt
package com.openland.spacex.store

import com.openland.spacex.InputValue
import org.json.JSONArray
import org.json.JSONObject

/**
 * Calculate Selector Key from JSON object
 * @param src Source Object
 * @return selector key
 */
private fun selectorValueKey(src: Any): String {
    if (src == JSONObject.NULL) {
        return "null"
    } else if (src is String) {
        return "\"" + src + "\""
    } else if (src is JSONObject) {
        val sortedKeys = src.keys().asSequence().toList().sorted()
        val values = mutableListOf<String>()
        for (k in sortedKeys) {
            values.add(k + ":" + selectorValueKey(src[k]))
        }
        return "{" + values.joinToString(",") + "}"
    } else if (src is JSONArray) {
        val res = mutableListOf<String>()
        for (i in 0 until src.length()) {
            res.add(selectorValueKey(src[i]))
        }
        return "[" + res.joinToString(",") + "]"
    } else {
        return src.toString()
    }
}

/**
 * Selector key from InputValue
 * @param value Input Value
 * @param arguments Operation arguments
 * @return selector key
 */
private fun selectorKey(value: InputValue, arguments: JSONObject): String? {
    if (value is InputValue.Int) {
        return value.value.toString()
    } else if (value is InputValue.Float) {
        return value.value.toString()
    } else if (value is InputValue.Boolean) {
        return value.value.toString()
    } else if (value is InputValue.String) {
        return "\"" + value.value + "\""
    } else if (value == InputValue.Null) {
        return "null"
    } else if (value is InputValue.List) {
        return "[" + value.items.map { selectorKey(it, arguments) }.joinToString(",") + "]"
    } else if (value is InputValue.Object) {
        val inner = value.fields
                .mapValues { selectorKey(it.value, arguments) }
                .toSortedMap()
                .filterValues { it != null }
                .map { it.key + ":" + it.value }
                .joinToString(",")
        return "{$inner}"
    } else if (value is InputValue.Reference) {
        return if (arguments.has(value.name)) {
            selectorValueKey(arguments.get(value.name))
        } else {
            null
        }
    }

    error("Unreachable code")
}

/**
 * Calculate field selector key
 * @param name Field Name
 * @param fieldArguments Field arguments
 * @param arguments Operation arguments
 * @return selector key
 */
fun selectorKey(name: String, fieldArguments: Map<String, InputValue>, arguments: JSONObject): String {
    if (fieldArguments.isEmpty()) {
        return name
    }
    val mapped = fieldArguments.mapValues { selectorKey(it.value, arguments) }
    val sortedKeys = mapped.keys.sortedBy { it }
    val converted = mutableListOf<String>()
    for (k in sortedKeys) {
        if (mapped[k] != null) {
            converted.add(k + ":" + mapped[k])
        }
    }
    if (converted.isEmpty()) {
        return name
    }
    return name + "(" + converted.joinToString(",") + ")"
}